public/challenge/index.js

```diff
@@ -0,0 +1,340 @@
+  $dateNowEndowed.checked = false;
+};
+
+const setDateNowEnabled = (endowed) => {
+  if (endowed) {
+    enableDateNow();
+  } else {
+    disableDateNow();
+  }
+};
+
+setDateNowEnabled(url.searchParams.get("dateNow") === "enabled");
+
+$dateNowEndowed.addEventListener("change", () => {
+  setDateNowEnabled($dateNowEndowed.checked);
+});
+
+// ********************
+// 2. We prepare APIs for the defender code.
+// ********************
+
+const getRandomValues = (array) => window.crypto.getRandomValues(array);
+
+// https://www.dafont.com/hyperspace.font
+const setMacguffinText = (text) => ($macguffin.innerText = text);
+
+/**
+ * @param {number} count
+ * @returns {void}
+ */
+const delayMS = (count) => {
+  // busywait for 'count' milliseconds
+  const target = Date.now() + count;
+  // eslint-disable-next-line no-empty
+  while (Date.now() < target) {}
+};
+
+// ********************
+// 3. We prepare the defender code.
+// ********************
+
+setMacguffinText("defender go");
+
+const buildSecretCode = () => {
+  const array = new Uint32Array(10);
+  let secretCode = "";
+  const SYMBOLS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  getRandomValues(array);
+  for (let i = 0; i < 10; i++) {
+    // we use Uint32Array, instead of Uint8Array, to reduce the bias
+    // somewhat. Actual launch codes should use a 128-bit integer before
+    // wrapping down to a single digit.
+    //
+    // Also please use don't use this for launch codes.
+    const digit = array[i] % SYMBOLS.length;
+    secretCode += SYMBOLS.slice(digit, digit + 1);
+  }
+  return secretCode;
+};
+
+const secretCode = buildSecretCode();
+setMacguffinText(secretCode);
+
+// ********************
+// 4. We prepare the confinement.
+// ********************
+
+let enableAttacker = false;
+const stopAttacker = () => {
+  enableAttacker = false;
+  $guessBox.classList.remove("attacking");
+};
+
+const startAttacker = () => {
+  enableAttacker = true;
+  $guessBox.classList.add("attacking");
+};
+
+const guess = (guessedCode) => {
+  // guess() is an exposed API: we must cast into a String to avoid TOCTOU
+  // (time-of-check time-of-use attacks).
+  guessedCode = `${guessedCode}`;
+
+  // To demonstrate how deterministic attacker code cannot sense covert
+  // channels, we provide a pretty obvious covert channel: we compare one
+  // character at a time, and busy-wait a long time between characters.
+  // The time we take indicates how many leading characters they got
+  // right, enabling a linear-time guessing attack.
+
+  $guess.innerText = guessedCode;
+
+  for (let i = 0; i < 10; i++) {
+    if (secretCode.slice(i, i + 1) !== guessedCode.slice(i, i + 1)) {
+      return false;
+    }
+    delayMS(10);
+  }
+
+  // they guessed correctly
+  stopAttacker();
+  return true;
+};
+
+const compartment = new Compartment(
+  harden({
+    console,
+    // See https://github.com/Agoric/agoric-sdk/issues/9515
+    assert: globalThis.assert,
+    guess,
+    ...endowments,
```

Comment:
This only builds the Compartment once per page-load, right? How does that work with `enableDateNow`/`disableDateNow` merely modifying `endowments` and changing the `window.history`, but not reloading the page? I haven't tried to interact with the demo to confirm, but it seems like the attacker's Compartment would continue to get the original authority.

This might require that `submitProgram` create a new Compartment each time.

Reply:
Indeed, I tested this by:
* load the challenge page
* click "Timing Channel Frustrated", which 1: loads the attack code, 2: leaves the "Expose" checkbox unchecked, 3: leaves the URL alone (no `?dateNow=enabled`)
* now click-on the Expose checkbox, which 1: checks the checkbox, 2: changes the URL to add `?dateNow=enabled`
* now click "Attack"
* observe: the attack fails (guesses all zeros and then repeats)

Reply:
Good catch.